File: PPC/Assignments/KnapsackProblem2/KnapsackProblemKotlin/src/main/kotlin/knapsack/channel/KnapsackGAChannel.kt
package knapsack.channel

import Individual
import Individual.Companion.deepCopy
import KnapsackGA
import KnapsackGA.Companion.N_GENERATIONS
import KnapsackGA.Companion.POP_SIZE
import KnapsackGA.Companion.PROB_MUTATION
import KnapsackGA.Companion.tournament
import knapsack.channel.Messages.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ThreadLocalRandom

class KnapsackGAChannel(
    override val silent: Boolean = false,
    private val nWorkers: Int = Runtime.getRuntime().availableProcessors(),
    private val chunkSize: Int = 10
) : KnapsackGA {
    private var population: Array<Individual> = Array(POP_SIZE) { Individual.createRandom(Random()) }

    override fun run(): Individual = runBlocking(Dispatchers.Default) {
        for (generation in 0 until N_GENERATIONS) {

            // Step1 - Calculate Fitness
            calculateFitness()

            // Step2 - Print the best individual so far.
            val best = bestOfPopulation()
            if (!silent) {
                println("KnapsackGAChannel: Best at generation $generation is $best with ${best.fitness}")
            }

            // Step3 - Find parents to mate (cross-over)
            val newPopulation = crossoverPopulation(best)

            // Step4 - Mutate
            mutatePopulation(newPopulation)

            population = newPopulation
        }

        return@runBlocking population.first()
    }

    private suspend fun calculateFitness() = coroutineScope {
        val messagePoolSize = (POP_SIZE + chunkSize - 1) / chunkSize
        val workChannel = Channel<ProcessIndividuals>(messagePoolSize)
        val resultChannel = Channel<ProcessedIndividuals>(messagePoolSize)
        val countDownLatch = CountDownLatch(nWorkers)

        repeat(nWorkers) {
            launch {
                for (message in workChannel) {
                    val subPopulation = message.population
                    for (individual in subPopulation) {
                        individual.measureFitness()
                    }
                    resultChannel.send(ProcessedIndividuals(message.idx, subPopulation))
                }
                countDownLatch.countDown()
            }
        }

        launch {
            countDownLatch.await()
            resultChannel.close()
        }

        for (idx in 0 until messagePoolSize) {
            val start = idx * chunkSize
            val end = minOf((idx + 1) * chunkSize, POP_SIZE)
            val chunk = population.sliceArray(start until end).deepCopy()

            workChannel.send(ProcessIndividuals(idx, chunk))
        }
        workChannel.close()

        for (result in resultChannel) {
            val startIdx = result.idx * chunkSize

            for (i in result.population.indices) {
                population[startIdx + i] = result.population[i]
            }
        }
    }

    private fun bestOfPopulation(): Individual {
        /*
		 * Returns the best individual of the population.
		 */
        return population.maxByOrNull { it.fitness } ?: population[0]
    }

    private suspend fun crossoverPopulation(best: Individual): Array<Individual> = coroutineScope {
        val toProcess = POP_SIZE - 1
        val messagePoolSize = (toProcess + chunkSize - 1) / chunkSize
        val workChannel = Channel<ProcessCrossoverIndividuals>(messagePoolSize)
        val resultChannel = Channel<ProcessedCrossoverIndividuals>(messagePoolSize)
        val countDownLatch = CountDownLatch(nWorkers)

        val newPopulation = Array(POP_SIZE) { best }

        repeat(nWorkers) {
            launch {
                for (message in workChannel) {
                    val population = message.population
                    val chunkSize = message.chunkSize

                    val newSubPopulation = Array(chunkSize) { Individual() }

                    val r = ThreadLocalRandom.current()

                    for (idx in 0 until chunkSize) {
                        // We select two parents, using a tournament.
                        val parent1 = tournament(r, population)
                        val parent2 = tournament(r, population)

                        newSubPopulation[idx] = parent1.crossoverWith(parent2, r)
                    }

                    resultChannel.send(ProcessedCrossoverIndividuals(message.idx, chunkSize, newSubPopulation))
                }
                countDownLatch.countDown()
            }
        }

        launch {
            countDownLatch.await()
            resultChannel.close()
        }

        val currentPop = population.deepCopy()

        for (idx in 0 until messagePoolSize) {
            val start = idx * chunkSize
            val remaining = toProcess - start
            val chunkSize = minOf(chunkSize, remaining)
            if (chunkSize <= 0) continue
            workChannel.send(
                ProcessCrossoverIndividuals(
                    idx,
                    chunkSize,
                    currentPop
                )
            )
        }
        workChannel.close()

        for (result in resultChannel) {
            val startIdx = 1 + result.idx * result.chunkSize

            for (i in result.newPopulation.indices) {
                newPopulation[startIdx + i] = result.newPopulation[i]
            }
        }

        return@coroutineScope newPopulation
    }

    private suspend fun mutatePopulation(newPopulation: Array<Individual>) = coroutineScope {
        val toProcess = POP_SIZE - 1
        val messagePoolSize = (toProcess + chunkSize - 1) / chunkSize
        val workChannel = Channel<ProcessIndividuals>(messagePoolSize)
        val resultChannel = Channel<ProcessedIndividuals>(messagePoolSize)
        val countDownLatch = CountDownLatch(nWorkers)

        repeat(nWorkers) {
            launch {
                for (message in workChannel) {
                    val subPopulation = message.population
                    val r = ThreadLocalRandom.current()
                    for (individual in subPopulation) {
                        if (r.nextDouble() < PROB_MUTATION) {
                            individual.mutate(r)
                        }
                    }
                    resultChannel.send(ProcessedIndividuals(message.idx, subPopulation))
                }
                countDownLatch.countDown()
            }
        }

        launch {
            countDownLatch.await()
            resultChannel.close()
        }

        for (idx in 0 until messagePoolSize) {
            val startIdx = 1 + idx * chunkSize
            val endIdx = minOf((idx + 1) * chunkSize, POP_SIZE)
            if (startIdx >= endIdx) continue

            val chunk = newPopulation.sliceArray(startIdx until endIdx).deepCopy()

            workChannel.send(ProcessIndividuals(idx, chunk))
        }
        workChannel.close()

        for (result in resultChannel) {
            val startIdx = 1 + result.idx * chunkSize
            for (i in result.population.indices) {
                newPopulation[startIdx + i] = result.population[i]
            }
        }

    }

}

File: PPC/Assignments/KnapsackProblem2/KnapsackProblemGo/knapsack/channel/knapsackGAChannel.go
package channel

import (
	"KnapsackProblemGo/knapsack"
	"fmt"
	"math/rand"
	"sync"
	"time"
)

type KnapsackChannel struct {
	population []*knapsack.Individual
	nWorkers   int
	chunkSize  int
}

type ProcessChunk struct {
	startIdx int
	endIdx   int
}

func NewKnapsackGAChannel(nWorkers int, chunkSize int) *KnapsackChannel {
	ga := &KnapsackChannel{
		population: make([]*knapsack.Individual, knapsack.PopSize),
		nWorkers:   nWorkers,
		chunkSize:  chunkSize,
	}

	ga.populateInitialPopulationRandomly()
	return ga
}

func (ga *KnapsackChannel) populateInitialPopulationRandomly() {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	for i := 0; i < knapsack.PopSize; i++ {
		ga.population[i] = knapsack.NewIndividualRandom(r)
	}
}

func (ga *KnapsackChannel) Run(silent bool) *knapsack.Individual {
	for g := 0; g < knapsack.NGenerations; g++ {
		// Step1 - Calculate Fitness
		ga.calculateFitness()

		// Step2 - Print the best individual so far.
		best := ga.bestOfPopulation()
		if !silent {
			fmt.Printf("KnapsackGAChannel: Best at generation %d has fitness %d\n", g, best.Fitness)
		}

		// Step3 - Find parents to mate (cross-over)
		newPopulation := ga.crossoverPopulation(best)

		// Step4 - Mutate
		ga.mutatePopulation(newPopulation)

		ga.population = newPopulation
	}

	return ga.population[0]
}

func (ga *KnapsackChannel) calculateFitness() {
	messagePoolSize := (knapsack.PopSize + ga.chunkSize - 1) / ga.chunkSize
	workChannel := make(chan ProcessIndividuals, messagePoolSize)
	resultChannel := make(chan ProcessedIndividuals, messagePoolSize)

	var wg sync.WaitGroup
	for i := 0; i < ga.nWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for message := range workChannel {
				chunk := message.population
				for _, ind := range chunk {
					ind.MeasureFitness()
				}
				resultChannel <- ProcessedIndividuals{
					idx:        message.idx,
					population: chunk,
				}
			}
		}()
	}

	go func() {
		wg.Wait()
		close(resultChannel)
	}()

	for i := 0; i < messagePoolSize; i++ {
		start := i * ga.chunkSize
		end := start + ga.chunkSize
		if end > knapsack.PopSize {
			end = knapsack.PopSize
		}
		chunk := knapsack.DeepCopy(ga.population[start:end])

		workChannel <- ProcessIndividuals{
			idx:        i,
			population: chunk,
		}
	}

	close(workChannel)

	for result := range resultChannel {
		start := result.idx * ga.chunkSize

		for j, ind := range result.population {
			ga.population[start+j] = ind
		}
	}
}

func (ga *KnapsackChannel) bestOfPopulation() *knapsack.Individual {
	best := ga.population[0]
	for _, other := range ga.population {
		if other.Fitness > best.Fitness {
			best = other
		}
	}
	return best
}

func (ga *KnapsackChannel) crossoverPopulation(best *knapsack.Individual) []*knapsack.Individual {
	toProcess := knapsack.PopSize - 1
	messagePoolSize := (toProcess + ga.chunkSize - 1) / ga.chunkSize
	workChannel := make(chan ProcessCrossoverIndividuals, messagePoolSize)
	resultChannel := make(chan ProcessedCrossoverIndividuals, messagePoolSize)

	newPopulation := make([]*knapsack.Individual, knapsack.PopSize)
	newPopulation[0] = best

	var wg sync.WaitGroup
	for i := 0; i < ga.nWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for message := range workChannel {
				idx := message.idx
				population := message.population
				chunkSize := message.chunkSize

				newSubPopulation := make([]*knapsack.Individual, chunkSize)
				r := rand.New(rand.NewSource(time.Now().UnixNano() + int64(idx)))

				for j := 0; j < chunkSize; j++ {
					parent1 := knapsack.Tournament(r, population)
					parent2 := knapsack.Tournament(r, population)

					newSubPopulation[j] = parent1.CrossoverWith(parent2, r)
				}

				resultChannel <- ProcessedCrossoverIndividuals{
					idx:           message.idx,
					chunkSize:     chunkSize,
					newPopulation: newSubPopulation,
				}
			}
		}()
	}

	go func() {
		wg.Wait()
		close(resultChannel)
	}()

	currentPop := knapsack.DeepCopy(ga.population)

	for i := 0; i < messagePoolSize; i++ {
		start := i * ga.chunkSize
		remaining := toProcess - start
		chunkSize := ga.chunkSize
		if remaining < ga.chunkSize {
			chunkSize = remaining
		}
		if chunkSize > 0 {
			workChannel <- ProcessCrossoverIndividuals{
				idx:        i,
				population: currentPop,
				chunkSize:  chunkSize,
			}
		}
	}
	close(workChannel)

	for result := range resultChannel {
		start := result.idx*ga.chunkSize + 1

		for j, ind := range result.newPopulation {
			newPopulation[start+j] = ind
		}
	}

	return newPopulation
}

func (ga *KnapsackChannel) mutatePopulation(newPopulation []*knapsack.Individual) {
	toProcess := knapsack.PopSize - 1
	messagePoolSize := (toProcess + ga.chunkSize - 1) / ga.chunkSize
	workChannel := make(chan ProcessIndividuals, messagePoolSize)
	resultChannel := make(chan ProcessedIndividuals, messagePoolSize)

	var wg sync.WaitGroup
	for i := 0; i < ga.nWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for message := range workChannel {
				idx := message.idx
				chunk := message.population
				r := rand.New(rand.NewSource(time.Now().UnixNano() + int64(idx)))
				for _, ind := range chunk {
					if r.Float64() < knapsack.ProbMutation {
						ind.Mutate(r)
					}
				}
				resultChannel <- ProcessedIndividuals{
					idx:        idx,
					population: chunk,
				}
			}
		}()
	}

	go func() {
		wg.Wait()
		close(resultChannel)
	}()

	for i := 0; i < messagePoolSize; i++ {
		start := i*ga.chunkSize + 1
		end := start + ga.chunkSize
		if end > knapsack.PopSize {
			end = knapsack.PopSize
		}

		chunk := knapsack.DeepCopy(newPopulation[start:end])
		workChannel <- ProcessIndividuals{
			idx:        i,
			population: chunk,
		}
	}
	close(workChannel)

	for result := range resultChannel {
		start := result.idx*ga.chunkSize + 1
		for j, ind := range result.population {
			newPopulation[start+j] = ind
		}
	}
}
